File: Interfaces/IUnitOfWork.cs
using cms_webapi.Models;

namespace cms_webapi.Interfaces
{
    /// <summary>
    /// Unit of Work pattern interface for managing transactions and repositories
    /// </summary>
    public interface IUnitOfWork : IDisposable
    {
        /// <summary>
        /// User repository
        /// </summary>
        IGenericRepository<User> Users { get; }
        IGenericRepository<Country> Countries { get; }
        IGenericRepository<City> Cities { get; }
        IGenericRepository<District> Districts { get; }
        IGenericRepository<CustomerType> CustomerTypes { get; }
        IGenericRepository<Customer> Customers { get; }
        IGenericRepository<Title> Titles { get; }
        IGenericRepository<Contact> Contacts { get; }
        IGenericRepository<Activity> Activities { get; }
        IGenericRepository<ProductPricing> ProductPricings { get; }
        IGenericRepository<ProductPricingGroupBy> ProductPricingGroupBys { get; }
        IGenericRepository<UserDiscountLimit> UserDiscountLimits { get; }
        IGenericRepository<PaymentType> PaymentTypes { get; }
        IGenericRepository<ShippingAddress> ShippingAddresses { get; }
        IGenericRepository<Quotation> Quotations { get; }



        /// <summary>
        /// Save all changes to the database within a transaction
        /// </summary>
        /// <returns>Number of affected rows</returns>
        Task<int> SaveChangesAsync();

        /// <summary>
        /// Begin a new transaction
        /// </summary>
        Task BeginTransactionAsync();

        /// <summary>
        /// Commit the current transaction
        /// </summary>
        Task CommitTransactionAsync();

        /// <summary>
        /// Rollback the current transaction
        /// </summary>
        Task RollbackTransactionAsync();

        /// <summary>
        /// Get repository for any entity type
        /// </summary>
        /// <typeparam name="T">Entity type</typeparam>
        /// <returns>Generic repository for the entity</returns>
        IGenericRepository<T> Repository<T>() where T : BaseEntity;
    }
}

File: WebApi/Interfaces/IUnitOfWork.ts
import { IGenericRepository } from './IGenericRepository';
import { BaseEntity } from '../Models/BaseEntity';
import { User } from '../Models/User';
import { Country } from '../Models/Country';
import { City } from '../Models/City';
import { District } from '../Models/District';
import { CustomerType } from '../Models/CustomerType';
import { Customer } from '../Models/Customer';
import { Title } from '../Models/Title';
import { Contact } from '../Models/Contact';
import { Activity } from '../Models/Activity';
import { ProductPricing } from '../Models/ProductPricing';
import { ProductPricingGroupBy } from '../Models/ProductPricingGroupBy';
import { UserDiscountLimit } from '../Models/UserDiscountLimit';
import { PaymentType } from '../Models/PaymentType';
import { ShippingAddress } from '../Models/ShippingAddress';
import { Quotation } from '../Models/Quotation';

export interface IUnitOfWork {
  users: IGenericRepository<User>;
  countries: IGenericRepository<Country>;
  cities: IGenericRepository<City>;
  districts: IGenericRepository<District>;
  customerTypes: IGenericRepository<CustomerType>;
  customers: IGenericRepository<Customer>;
  titles: IGenericRepository<Title>;
  contacts: IGenericRepository<Contact>;
  activities: IGenericRepository<Activity>;
  productPricings: IGenericRepository<ProductPricing>;
  productPricingGroupBys: IGenericRepository<ProductPricingGroupBy>;
  userDiscountLimits: IGenericRepository<UserDiscountLimit>;
  paymentTypes: IGenericRepository<PaymentType>;
  shippingAddresses: IGenericRepository<ShippingAddress>;
  quotations: IGenericRepository<Quotation>;

  saveChanges(): Promise<number>;
  beginTransaction(): Promise<void>;
  commitTransaction(): Promise<void>;
  rollbackTransaction(): Promise<void>;
  repository<T extends BaseEntity>(): IGenericRepository<T>;
}